src/backend/distributed/utils/maintenanced.c

```diff
@@ -292,13 +292,35 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		proc_exit(0);
 	}
 
+	if (myDbData->workerPid != 0 && myDbData->workerPid != MyProcPid)
+	{
+		/*
+		 * Another maintenance daemon is running. This usually happens because
+		 * postgres restarts the daemon after an non-zero exit, and
+		 * InitializeMaintenanceDaemonBackend started one before postgres did.
+		 * In that case, the first one stays and the last one exits.
+		 */
+
+		proc_exit(0);
+	}
```

Comment:
There is an assert with a pid check in the main loop as well - without acquiring a lock and relookup though.

Would it make sense to verify every loop rotation that we are still the designated maintenance worker for this database?

I don't think it can ever happen with the locks acquired that the pid might change while it is running. But yeah, we also missed this current case. Extra defence here might help. I don't know how intrusive the lock acquisition and lookup would be, but during normal operation the maintenance daemon should not have to do much. Only if the lock would require all running backends to synchronise I would think we can omit the extra check.

Reply:
I did some additional clean up to move this Assert into MaintenanceDaemonShmemExit